File: 05_hash_tables/Haskell/02_check_voter.hs
import qualified Data.HashMap.Strict as Map

all_voters = Map.fromList[("tom", True), ("mike", True)]

check_voter name voters_map = case Map.lookup name voters_map of
  Just _ -> ("kick them out!", voters_map)
  otherwise -> ("let them vote!", Map.insert name True voters_map)

log_vote name voters_map = do
  putStrLn str
  return new_voters_map
  where (str, new_voters_map) = check_voter name voters_map

main = do
  try_mike <- log_vote "mike" all_voters
  try_john <- log_vote "john" try_mike
  return try_john

File: 05_hash_tables/ES6/02_check_voter.js
const voted = {};
const checkVoter = (name) => {
  if (voted[name]) {
    console.log('kick them out!');
  } else {
    voted[name] = true;
    console.log('let them vote!');
  }
};


checkVoter('tom'); // let them vote!
checkVoter('mike'); // let them vote!
checkVoter('mike'); // kick them out!
